### PR TITLE
[WIP][DONOTMERGE]: add Endpoint to be used for setting up ingress network for PVC data transfer

### PR DESCRIPTION
**Pull request by alaypatel07**
**Describe what this PR does**
This PR creates an endpoint package for a collection of implementations that can be used to allow traffic into a namespace of a cluster for PVC data transfer.

This will be followed up by other packages, a transport (tunneling implementations) and a transfer (implementation for PVC data servers and client). The overall usage is in [here](https://github.com/alaypatel07/volsync/pull/1)

This PR has an example commit which has to be removed before merging.

### CODE REVIEW

File: lib/endpoint/route/route.go [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+							IntVal: port,
+						},
+					},
+				},
+				Selector: serviceSelector,
+				Type:     corev1.ServiceTypeClusterIP,
+			}
+		}
+
+		service.Labels = r.objMeta.Labels()
```
Comment (JohnStrunk) [inline lib/endpoint/route/route.go:211]: This is the same as `serviceSelector`. Why retrieve it again?

**Review by JohnStrunk — COMMENTED** [review]
General thoughts:

I don't really understand the purpose of the `lib/meta` directory. Is the idea that it's a way to generate a list of changes that can be applied at some time later? If so, I would have expected to see it structured as a set of mutation functions, similar to controller-runtime createOrUpdate().

The label manipulation makes me a bit nervous because it's overwriting labels (for the Service, in particular) as opposed to upserting. If another controller mutates the labels, there can be problems. We ran into this w/ submariner and globalnet, though the specific case may have been annotations.

File: main.go [hunk 2]
```diff
@@ -101,6 +102,12 @@ func main() {
 		os.Exit(1)
 	}
 
+	_, err = route.APIsToWatch(mgr.GetClient())
```
[the remark this answers is not in the document]
Comment (alaypatel07) [reply main.go:105]: >  We need the behavior to be that the functionality is just disabled.

I agree, that is the eventual goal. will refactor with that in mind

File: lib/endpoint/endpoint.go [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+	// Hostname returns a hostname for the endpoint
+	Hostname() string
+	// BackendPort returns a port which can be used by the application behind this endpoint
+	BackendPort() int32
+	// IngressPort is a port which is used by the clients to connect to the endpoint
+	IngressPort() int32
+	// IsHealthy returns whether or not all Kube resources used by endpoint are healthy
+	IsHealthy(c client.Client) (bool, error)
+	// MarkForCleanup
+	MarkForCleanup(c client.Client, key, value string) error
```
[the remark this answers is not in the document]
Comment (alaypatel07) [reply lib/endpoint/endpoint.go:21]: +1. I will refactor

File: lib/endpoint/route/route.go [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+	service := &corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      r.NamespacedName().Name,
+			Namespace: r.NamespacedName().Namespace,
+		},
+	}
+
+	// TODO: log the return operation from CreateOrUpdate
+	_, err := controllerutil.CreateOrUpdate(context.TODO(), c, service, func() error {
+		if service.CreationTimestamp.IsZero() {
```
[the remark this answers is not in the document]
Comment (alaypatel07) [reply lib/endpoint/route/route.go:193]: +1, I think the immutable fiends(spec.ClusterIP and spec.ClusterIPs are not being used here, will refactorr

**Pull request closed without merging**